### PR TITLE
Cleared out Cloudant, CouchDB and GSON dependencies in core

### PR DESCRIPTION
**Pull request by psuter**

### CODE REVIEW

**Comment by rabbah** [comment]
PG?

File: common/scala/src/whisk/core/database/CouchDbRestStore.scala [hunk 1]
```diff
@@ -46,74 +46,58 @@ import whisk.common.LoggingMarkers
...
 
-        // We'll activate this warning once we start getting rid of GSON.
-        // It doesn't really matter for now, because we insert from the GSON view
-        // anyway.
-        if (false && asJson != asGson) {
-            warn(this, s"[PUT] JSON/GSON mismatch:")
-            warn(this, s"[PUT]   - json : " + asJson)
-            warn(this, s"[PUT]   - gson : " + asGson)
-        }
+        val id: String = asJson.fields("_id").convertTo[String].trim
```
Comment (rabbah) [inline common/scala/src/whisk/core/database/CouchDbRestStore.scala]: is `_id` guaranteed to exist? could this throw an exception before the require below?

File: common/scala/src/whisk/core/database/CouchDbRestStore.scala [hunk 2]
```diff
@@ -135,7 +119,7 @@ class CouchDbRestStore[Unused, DocumentAbstraction <: DocumentSerializer](
                 case Left(StatusCodes.NotFound) =>
                     info(this, s"[DEL] '$dbName', document: '${doc}'; not found.", LoggingMarkers.DATABASE_DELETE_DONE)
                     // for compatibility
-                    throw new org.lightcouch.NoDocumentException("not found on 'del'")
+                    throw NoDocumentException("not found on 'del'")
```
Comment (rabbah) [inline common/scala/src/whisk/core/database/CouchDbRestStore.scala]: del -> delete in exception message?

File: common/scala/src/whisk/core/database/CouchDbRestStore.scala [hunk 3]
```diff
@@ -250,7 +231,11 @@ class CouchDbRestStore[Unused, DocumentAbstraction <: DocumentSerializer](
     }
 
     private def reportFailure[T, U](f: Future[T], onFailure: Throwable => U): Future[T] = {
-        f.onFailure({ case x => onFailure(x) })
+        f.onFailure({
+            // These failures are intentional and shouldn't trigger the catcher.
+            case _ : ArtifactStoreException => ;
```
Comment (rabbah) [inline common/scala/src/whisk/core/database/CouchDbRestStore.scala]: semicolon?! the horror.

File: common/scala/src/whisk/core/database/DocumentFactory.scala [hunk 4]
```diff
@@ -176,7 +169,7 @@ trait DocumentFactory[R, W] extends (R => Try[W]) with InMemoryCache[R, W] {
     }
 
     /**
-     * Fetches a raw record of type R from the datastore by its id (and revision if given)
+     * FIXME UPDATE Fetches a raw record of type R from the datastore by its id (and revision if given)
```
Comment (rabbah) [inline common/scala/src/whisk/core/database/DocumentFactory.scala]: assuming the fixme will be resolved as part of this pr...

File: common/scala/src/whisk/core/entity/WhiskStore.scala [hunk 5]
```diff
@@ -72,23 +68,36 @@ protected[core] trait WhiskDocument
...
+     * differ from `toJson` in exceptional cases.
+     */
+    override def toDocumentRecord: JsObject = {
+        val id = docid.id
+        val revOrNull = rev.rev
+
+        // Building up the fields.
+        val base    = this.toJson.fields
+        val withId  = base + ("_id" -> JsString(id))
+        val withRev = if(revOrNull == null) withId else { withId + ("_rev" -> JsString(revOrNull)) }
```
Comment (rabbah) [inline common/scala/src/whisk/core/entity/WhiskStore.scala]: space after if?

File: common/scala/src/whisk/core/database/CouchDbRestStore.scala (hunk 1 above)
Comment (psuter) [inline common/scala/src/whisk/core/database/CouchDbRestStore.scala]: `.toDocumentRecord` in `WhiskDocument` adds it

File: common/scala/src/whisk/core/database/DocumentFactory.scala (hunk 4 above)
Comment (psuter) [inline common/scala/src/whisk/core/database/DocumentFactory.scala]: Turns out making classes immutable is an orthogonal question (although now it's at least possible). I can still do it as part of this PR.

**Comment by psuter** [comment]
Perf PG 562

**Comment by rabbah** [comment]
Passed PG 344.

**Comment by rabbah** [comment]
Rebased and merged as https://github.com/openwhisk/openwhisk/compare/55e369778802...53e11eb87bba.

**Pull request closed without merging**